### PR TITLE
Bach BQ: Removed `materialization` when in DataFrame.groupby

### PR DESCRIPTION
**Pull request by KathiaBarahona**
Decided to remove the automatic materialization  for BQ when grouping by as it was a bit annoying when applying window functions, since `df.groupby` removes sorting for the df. Also it is a bit frustrating when trying to do simple operations and ending up with an unnecessary subquery. After playing around a bit more with this, I noticed that we can reference the expression used in the `groupby` by just referencing to the column from the select:

For example, for the following data:
```sql
WITH PlayerStats AS
 (
  SELECT 'Adams' as LastName, 51 as OpponentID, 3 as PointsScored UNION ALL
  SELECT 'Buchanan', 77, 0 UNION ALL
  SELECT 'Coolidge', 77, 1 UNION ALL
  SELECT 'Adams', 52, 4 UNION ALL
  SELECT 'Buchanan', 50, 13 UNION ALL
  SELECT 'Barahona', 50, -5
)
```
Making reference to the same expression in both SELECT  and GROUP BY clause will be something like:
```
SELECT 
LEFT(`LastName`, 2),
SUM(PointsScored) AS `TotalPointsScored`
FROM PlayerStats
GROUP BY LEFT(`LastName`, 2)
```
Now, this introduces many conflicts when adding the column reference in the select expression `e.g LEFT(LastName, 2) as LastName` and considering the same expression in the GROUP BY clause will actually do something like:
`LEFT(LEFT(LastName, 2), 2)` which is incorrect (**this is not the case for Postgres, it will actually consider the main column**). Now, the solution for this is actually just making reference to the alias in GROUP BY  clause for BigQuery, resulting into:

```
SELECT 
LEFT(`LastName`, 2) AS `LastName`,
SUM(PointsScored) AS `TotalPointsScored`
FROM PlayerStats
GROUP BY `LastName`
```

After removing the materialization, I noticed some problems regards the `ORDER BY`. Seems that the ORDER BY clause works similar to the GROUP BY.
Therefore, I think that if ORDER BY expression diverge from GROUP BY  or AGGREGATED expressions, we should raise an exception and tell the user to remove or changed the sort for the DataFrame. As it is way too difficult to know what the user wants to actually sort by

### CODE REVIEW

**Review by thijs-obj — APPROVED** [review]
Quite complex stuff to review. Quite happy we have a test suite to give some confidence in correctness, cause I find it hard to make sure all cases are covered while reviewing. 

Left some small comments, but mostly looks good to me. Fine to merge and fix stuff up later today also if this blocks anything

File: bach/bach/dataframe.py [hunk 1]
```diff
@@ -1983,27 +1946,44 @@ def _get_order_by_clause(self) -> Expression:
...
                 exprs.extend(multi_lvl_exprs)
                 fmtstr.extend([fmt] * len(multi_lvl_exprs))
+                continue
 
-            return Expression.construct(f'order by {", ".join(fmtstr)}', *exprs)
-        else:
-            return Expression.construct('')
+            expr = sc.expression
+            if self.group_by and is_bigquery(self.engine):
+                expr = Expression.column_reference(all_series_expr[expr.to_sql(dialect)])
```
[the remark this answers is not in the document]
Comment (KathiaBarahona) [reply bach/bach/dataframe.py:1982]: mmmmmmm probably yes, but thats a veery extreme edge.. like the user created a multi-level series manually and did random stuff with the expressions. Tho, I think we can ignore it for now. Here the issue to consider this https://github.com/objectiv/objectiv-analytics/issues/943

File: bach/bach/dataframe.py [hunk 2]
```diff
@@ -1983,27 +1946,44 @@ def _get_order_by_clause(self) -> Expression:
         Get a properly formatted order by expression based on this df's order_by.
         Will return an empty Expression in case ordering is not requested.
         """
-        if self._order_by:
-            exprs = []
-            fmtstr = []
+        if not self._order_by:
```
[the remark this answers is not in the document]
Comment (KathiaBarahona) [reply bach/bach/dataframe.py:1949]: Fixed here https://github.com/objectiv/objectiv-analytics/pull/938/commits/51d89e81d8c5ef139bfe02d675e18dd4b1d51239

File: bach/bach/dataframe.py [hunk 3]
```diff
@@ -1983,27 +1946,44 @@ def _get_order_by_clause(self) -> Expression:
...
+        if (
+            self.group_by and
+            not all(
+                ob.expression.to_sql(dialect) in all_series_expr for ob in self._order_by
+            )
+        ):
+            raise Exception(
+                'Current order by clause is referencing expressions that are neither '
+                'aggregated or grouped. Please call DataFrame.sort_value or DataFrame.sort_index '
+                'and try materializing again.'
```
[the remark this answers is not in the document]
Comment (KathiaBarahona) [reply bach/bach/dataframe.py]: changed the message here https://github.com/objectiv/objectiv-analytics/pull/938/commits/afb66ef6fdd8712743e269f68384eacfb36870c0

**Pull request merged**